Add numpy dependency to python wrapper.

404Vector opened the pull request:
There was an issue with the existing [PR](https://github.com/AcademySoftwareFoundation/openexr/pull/1933), so I'm submitting a separate PR that resolves the Numpy dependency issue.


- Modify pyproject.toml.
Add numpy to package dependencies. When installing this package with pip, numpy will be installed additionally if it is not in the Python environment.

- Modify python-related workflows.
Modify each python workflow to explicitly install the dependencies (pyproject.toml) required for the build.

close #1919

404Vector committed:
update pyproject.toml and python workflows: Add numpy dependency

- Additional dependencies required for the build environment
Numpy must be installed in the python environment where the build is being performed in order to build.

- Automatic installation of dependencies like a normal python package
When installing this package, if numpy does not exist in the python environment, it will be installed automatically.

Signed-off-by: HyeongSeok Kim <tiryul@gmail.com>

JeanChristopheMorinPerso commented:
Can I ask what's the rationale for adding a dependency on numpy? Technically someone could use any other libraries that have a compatible numpy like array types.

JeanChristopheMorinPerso commented on .github/workflows/python-wheels-publish-test.yml:
```diff
@@ -62,6 +62,10 @@ jobs:
           CIBW_TEST_SKIP: "*-macosx_universal2:arm64"
           CIBW_ENVIRONMENT: OPENEXR_RELEASE_CANDIDATE_TAG="${{ github.ref_name }}"
+          CIBW_BEFORE_BUILD: |
+            pip install tomli
```
This is not needed. The build frontend (pip, build, etc) will automatically install the dependencies specified in the pyproject's build-system.requires section.

JeanChristopheMorinPerso commented on pyproject.toml:
```diff
@@ -2,7 +2,7 @@
 
 [build-system]
-requires = ["scikit-build-core==0.10.7", "pybind11"]
+requires = ["scikit-build-core==0.10.7", "pybind11", "numpy"]
```
Numpy is not needed at build time. Pybind11 doesn't need numpy and will not use numpy.

JeanChristopheMorinPerso commented:
Actually, never mind we indeed need to add it as a runtime dependency because `pybind11::array` is used instead of `pybind11::buffer`. 

But it only needs `numpy>=1.7.0`.

>Note that pybind11/numpy.h does not depend on the NumPy headers, and thus can be used without declaring a build-time dependency on NumPy; NumPy>=1.7.0 is a runtime dependency.

https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html#arrays

But still, my comment about the build time dependency is still valid. We don't need numpy to build openexr.

JeanChristopheMorinPerso commented:
But I think I'll try to build locally and test locally, because I'd like to get a better traceback/backtrace.

404Vector committed:
update python-wheels-*yml, pyproject.toml: remove BEFORE_BUILD, remove numpy in build requires

- python-wheels-*yml: When testing on other PRs, I found that dependencies were not properly installed for each build case, so I added this code. However, this PR is for adding numpy dependencies, so it is not needed as @JeanChristopheMorinPerso suggested.

- pyproject.toml: If you are going to do any additional work with the build artifacts, like generating stubs, you will need numpy in your build environment, but it is not needed at the moment. I agree with the review by @JeanChristopheMorinPerso.

Signed-off-by: HyeongSeok Kim <tiryul@gmail.com>

404Vector commented:
> Can I ask what's the rationale for adding a dependency on numpy? Technically someone could use any other libraries that have a compatible numpy like array types.

@JeanChristopheMorinPerso, thank you for your review. I agree with what you said and have modified the code.

cary-ilm reviewed (approved):
Looks good to me, thanks. The CI check failure is because the workflow is out of date, already fixed on the main branch.
